jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/CachingHttpContentFactory.java

```diff
@@ -0,0 +1,571 @@
+    }
+
+    /**
+     * Tests whether the given HttpContent is cacheable, and if there is enough room to fit it in the cache.
+     *
+     * @param httpContent the HttpContent to test.
+     * @return whether the HttpContent is cacheable.
+     */
+    protected boolean isCacheable(HttpContent httpContent)
+    {
+        if (httpContent == null)
+            return true;
+
+        if (httpContent.getResource().isDirectory())
+            return false;
+
+        if (_maxCachedFiles <= 0)
+            return false;
+
+        // Will it fit in the cache?
+        long len = httpContent.getBytesOccupied();
+        return (len <= _maxCachedFileSize && len <= _maxCacheSize);
+    }
+
+    @Override
+    public HttpContent getContent(String path) throws IOException
+    {
+        CachingHttpContent cachingHttpContent = _cache.get(path);
+        if (cachingHttpContent != null)
+        {
+            cachingHttpContent.setLastAccessedNanos(NanoTime.now());
+            if (cachingHttpContent.isValid())
+            {
+                cachingHttpContent.retain();
+                return (cachingHttpContent instanceof NotFoundHttpContent) ? null : cachingHttpContent;
+            }
+            else
+                removeFromCache(cachingHttpContent);
+        }
+
+        HttpContent httpContent = _authority.getContent(path);
+        if (isCacheable(httpContent))
+        {
+            AtomicBoolean wasAdded = new AtomicBoolean(false);
+            cachingHttpContent = _cache.computeIfAbsent(path, p ->
+            {
+                wasAdded.set(true);
+                CachingHttpContent cachingContent = (httpContent == null) ? newNotFoundContent(p) : newCachedContent(p, httpContent);
+                _cachedSize.addAndGet(cachingContent.getBytesOccupied());
+                return cachingContent;
+            });
+
+            cachingHttpContent.retain();
+            if (wasAdded.get())
+                shrinkCache();
+            return (cachingHttpContent instanceof NotFoundHttpContent) ? null : cachingHttpContent;
+        }
+        return httpContent;
+    }
+
+    protected CachingHttpContent newCachedContent(String p, HttpContent httpContent)
+    {
+        return new CachedHttpContent(p, httpContent);
+    }
+
+    protected CachingHttpContent newNotFoundContent(String p)
+    {
+        return new NotFoundHttpContent(p);
+    }
+
+    protected interface CachingHttpContent extends HttpContent
+    {
+        long getLastAccessedNanos();
+
+        void setLastAccessedNanos(long nanosTime);
+
+        String getKey();
+
+        boolean isValid();
+
+        void retain();
+    }
+
+    protected class CachedHttpContent extends HttpContentWrapper implements CachingHttpContent
+    {
+        private final ByteBuffer _buffer;
+        private final String _cacheKey;
+        private final HttpField _etagField;
+        private final long _contentLengthValue;
+        private final AtomicLong _lastAccessed = new AtomicLong();
```

Comment:
Looks like a `volatile long` is all you need.

Reply:
done